### PR TITLE
feat: Select multiple rows and copy

### PR DESCRIPTION
**Pull request by elvus**
## Change Summary
Add new variables:
- selectedRows
- rangerows
- isSelectedBlock

Add new event listener:
- mousedown : on td (line 534)
- mouseover: on td (line 534)
- selectstart: on tbody (line 491)
- mouseup: on document (line 369)

Add new functions:
- selectBlock
- selectedRange
- startSelectRange
- clearRangeRows

Update:
- selectCell (function) add `clearRangeRows()`
- onKeyDown (function) add  `clearRangeRows()` and modify the default option (copy text)
- onClickOutside (function) add  `clearRangeRows()`
- Attr :class at line 540
```
active: isUIAllowed('xcDatatableEditable') && selected.col === colIndex && selected.row === rowIndex,
```
To
```
active: isUIAllowed('xcDatatableEditable') && selected.col === colIndex && selected.row === rowIndex || isUIAllowed('xcDatatableEditable') && selectedRange(rowIndex, colIndex),
```
## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Additional information / screenshots (optional)

https://user-images.githubusercontent.com/24416449/189783105-33a1a823-5e7b-4b74-b08f-d4c604f680c7.mp4

### CODE REVIEW

**Comment by o1lab** [comment]
@elvus : thank you for sending the PR. A few comments in the code would be helpful.

**Comment by bcakmakoglu** [comment]
As mentioned in the PR that you had closed:

Your code is unreadable to anyone except yourself, I can't follow the logic at all.
There are no comments to help understand what you built there or anything else to actually help other contributors follow your train of thought here.

Furthermore the code does not uphold *any* ESLint rules - *The rules don't exist to be ignored!*

If you need assistance, please say so and I will gladly help you improve this PR but in this state the PR cannot be accepted.

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 1]
```diff
@@ -398,7 +488,7 @@ const showContextMenu = (e: MouseEvent, target?: { row: number; col: number }) =
               </th>
             </tr>
           </thead>
-          <tbody>
+          <tbody v-on:selectstart.prevent>
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:491]: use `@selectstart` not `v-on`.

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 2]
```diff
@@ -283,10 +366,17 @@ const onKeyDown = async (e: KeyboardEvent) => {
 }
 
 useEventListener(document, 'keydown', onKeyDown)
+useEventListener(document, 'mouseup', (e: MouseEvent)=>{
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:369]: unnecessary event typing - `MouseEvent` can be removed.

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 3]
```diff
@@ -222,42 +285,62 @@ const onKeyDown = async (e: KeyboardEvent) => {
...
     case 'ArrowDown':
+      clearRangeRows()
       e.preventDefault()
       if (selected.row < data.value.length - 1) selected.row++
       break
     default:
       {
         const rowObj = data.value[selected.row]
         const columnObj = fields.value[selected.col]
+        let cptext = "";
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:324]: - what is `cptext`?
- single quotes
- no semi-colon at EOL

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 4]
```diff
@@ -222,42 +285,62 @@ const onKeyDown = async (e: KeyboardEvent) => {
...
+        if(rangerows.minrow !== null && rangerows.maxrow !== null && rangerows.mincol !== null && rangerows.maxcol !== null){
+          const cprows = data.value.slice(rangerows.minrow, rangerows.maxrow+1);
+          const cpcols = fields.value.slice(rangerows.mincol, rangerows.maxcol+1);
+          cprows.forEach((row : any)=>{
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:328]: why `any`?

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 5]
```diff
@@ -222,42 +285,62 @@ const onKeyDown = async (e: KeyboardEvent) => {
...
+            cpcols.forEach((col : any)=>{
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:329]: Why `any`?

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 6]
```diff
@@ -222,42 +285,62 @@ const onKeyDown = async (e: KeyboardEvent) => {
...
+              cptext = cptext + row.row[col.title]+'\t';
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:330]: No `+` for string concat, use interpolation

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 7]
```diff
@@ -222,42 +285,62 @@ const onKeyDown = async (e: KeyboardEvent) => {
...
+            })
+            cptext = cptext.trim()+'\n'
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:332]: No + for string concat, use interpolation

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 8]
```diff
@@ -198,11 +256,16 @@ const makeEditable = (row: Row, col: ColumnType) => {
 
 /** handle keypress events */
 const onKeyDown = async (e: KeyboardEvent) => {
+  if(selectedRows.startrow !== null && selectedRows.startcol !== null && selectedRows.endrow !== null && selectedRows.endcol !== null){
+    selected.row = selectedRows.startrow
+    selected.col = selectedRows.startcol
+  }
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:262]: Spacing

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 9]
```diff
@@ -56,8 +56,11 @@ const hasEditPermission = isUIAllowed('xcDatatableEditable')
 const isView = false
 
 const selected = reactive<{ row: number | null; col: number | null }>({ row: null, col: null })
+const selectedRows = reactive<{ startcol: number | null; endcol: number | null; startrow: number | null; endrow: number | null }>({ startcol: null, endcol: null, startrow: null, endrow: null })
+const rangerows = reactive<{ minrow: number | null; maxrow: number | null; mincol: number | null; maxcol: number | null }>({ mincol: null, maxcol: null, minrow: null, maxrow: null })
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:60]: - Why use `number | null` instead of just `number` and assign `NaN` as a falsy value and test against that?
That would make the Generic of `reactive` unnecessary and you could shorten this to
```ts
const selectedRows = reactive({ startcol: NaN, endcol: NaN, startrow: NaN, endrow: NaN })
```

- Use proper pascal casing, i.e. `startcol` should be `startCol`

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 10]
```diff
@@ -125,10 +128,65 @@ openNewRecordFormHook?.on(async () => {
...
+      selectedRows.endrow = row;
+    }
+  }else if(selected.col !== col || selected.row !== row){
+    if(isSelectedBlock){
+      selected.col = null;
+      selected.row = null;
+      selectedRows.endcol = col;
+      selectedRows.endrow = row;
+    }
+  }
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:149]: Why check `if (isSelectedBlock)` in each case instead of checking it before all cases?

```ts
if (isSelectedBlock) {
   if (selected.col === null || selected.row === null) { ... }
   else if ( ... ) { ... }
}

File: packages/nc-gui/components/smartsheet/Grid.vue [hunk 11]
```diff
@@ -125,10 +128,65 @@ openNewRecordFormHook?.on(async () => {
...
+const clearRangeRows = ()=>{
+  rangerows.mincol = null
+  rangerows.maxcol = null
+  rangerows.minrow = null
+  rangerows.maxrow = null
+  selectedRows.startrow = null
+  selectedRows.startcol = null 
+  selectedRows.endrow = null
+  selectedRows.endcol = null
+}
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Grid.vue:188]: - Can this whole block not be simplified?
- If not, add comments so others, especially other contributors, somehow can understand what you meant to do here. Otherwise nobody except yourself can debug this if there are issues
- Possibly outsource this to somewhere else because the Grid.vue file is getting bloated
- Spacing
- No semi colon on EOL

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

**Comment by elvus** [comment]
Hi @bcakmakoglu, can i contact whit you on discord?

**Comment by bcakmakoglu** [comment]
@elvus Sure :) You can find me in the Noco Discord as `Braks`

**Comment by psytron** [comment]
That feature demo video is top class. Really excited for this key feature.

**Comment by emdax** [comment]
> ## Change Summary
> Add new variables:
> 
> * selectedRows
> * rangerows
> * isSelectedBlock
> 
> Add new event listener:
> 
> * mousedown : on td (line 534)
> * mouseover: on td (line 534)
> * selectstart: on tbody (line 491)
> * mouseup: on document (line 369)
> 
> Add new functions:
> 
> * selectBlock
> * selectedRange
> * startSelectRange
> * clearRangeRows
> 
> Update:
> 
> * selectCell (function) add `clearRangeRows()`
> * onKeyDown (function) add  `clearRangeRows()` and modify the default option (copy text)
> * onClickOutside (function) add  `clearRangeRows()`
> * Attr :class at line 540
> 
> ```
> active: isUIAllowed('xcDatatableEditable') && selected.col === colIndex && selected.row === rowIndex,
> ```
> 
> To
> 
> ```
> active: isUIAllowed('xcDatatableEditable') && selected.col === colIndex && selected.row === rowIndex || isUIAllowed('xcDatatableEditable') && selectedRange(rowIndex, colIndex),
> ```
> 
> ## Change type
> * [x]  feat: (new feature for the user, not a new feature for build script)
> * [ ]  fix: (bug fix for the user, not a fix to a build script)
> * [ ]  docs: (changes to the documentation)
> * [ ]  style: (formatting, missing semi colons, etc; no production code change)
> * [ ]  refactor: (refactoring production code, eg. renaming a variable)
> * [ ]  test: (adding missing tests, refactoring tests; no production code change)
> * [ ]  chore: (updating grunt tasks etc; no production code change)
> 
> ## Additional information / screenshots (optional)
>  nocodb-mozilla-firefox-2022-09-12-20-32-22_HECyWDy4.mp4
Marvelous feature preview. Glorious functionality.

**Comment by bcakmakoglu** [comment]
I'll be looking into this today, so you'll get your hands on this feature soon :)
